### PR TITLE
EM-835: Special Character Fixes Part Deux

### PR DESCRIPTION
**Pull request by toastercup**
* Ensure special characters in URI path are encoded for requests via new `NormalizeURIPath` middleware
* Refactor specs using WebMock to accurately test this use case
* Allow semantic versioning of `oauth2` gem

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,9 @@
+root = true
```
Comment (Casao) [inline .editorconfig]: Good file, but I would put it in its own PR just for single purpose PRs (just my opinion)

File: lib/cortex/client.rb [hunk 2]
```diff
@@ -20,7 +20,7 @@ class Client
     include Cortex::Request
 
     def initialize(hasharg)
-      @base_url = hasharg[:base_url] || 'https://cbcortex.com/api/v1'
+      @base_url = hasharg[:base_url] || 'http://cortex.dev/api/v1'
```
Comment (Casao) [inline lib/cortex/client.rb]: I do not like this -- we go from defaulting to prod to defaulting to dev, potentially breaking a lot of use cases. I would say the current version should check if it exists, drop some warning if not, and still default to prod. Next version can make the change without destroying anything.

File: lib/cortex/faraday_middleware/encode_uri_path.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+require 'faraday/middleware'
+require 'addressable/uri'
+
+module Cortex
+  module FaradayMiddleware
+    class EncodeURIPath < Faraday::Middleware
+      def call(env)
+        env[:url].path = Addressable::URI.encode(env[:url].path)
```
Comment (Casao) [inline lib/cortex/faraday_middleware/encode_uri_path.rb]: Blindly encoding sets you up for double encoding -- example:

```rb
[10] pry(main)> x = Addressable::URI.parse('http://www.careerbuilder.com/as df?sadfsa=tes$t')
=> #<Addressable::URI:0x3fbfd66941d4 URI:http://www.careerbuilder.com/as df?sadfsa=tes$t>
[11] pry(main)> x.path
=> "/as df"
[12] pry(main)> Addressable::URI.encode(x.path)
=> "/as%20df"
[13] pry(main)> Addressable::URI.encode(Addressable::URI.encode(x.path))
=> "/as%2520df"
```

I think noramlize_component might be better? https://github.com/sporkmonger/addressable/blob/master/lib/addressable/uri.rb#L461

File: spec/posts_spec.rb [hunk 4]
```diff
@@ -18,8 +16,21 @@
       expect(client.posts.get_published(1)).to eq('response')
     end
 
-    it 'should work with special characters' do
-      expect { client.posts.get_published('1 post') }.to_not raise_error(URI::InvalidURIError)
+    context 'with special characters' do
+      before(:all) do
+        stub_request(:get, 'http://cortex.dev/api/v1/posts/feed/1%20post?access_token=123')
```
Comment (Casao) [inline spec/posts_spec.rb]: instead of this, try:

```rb
let!(:stubbed_request) { stub_request(:get, 'http://cortex.dev/api/v1/posts/feed/1%20post?access_token=123') }

expect(stubbed_request).to have_been_made.once
```

The ! on let ensures it stubs, and storing a reference means you don't have to change the stub twice if you change your pattern for some reason.

File: lib/cortex/client.rb (hunk 2 above)
Comment (toastercup) [inline lib/cortex/client.rb]: In this case, the original URL was invalid - it doesn't actually resolve to anything

File: lib/cortex/faraday_middleware/encode_uri_path.rb (hunk 3 above)
Comment (toastercup) [inline lib/cortex/faraday_middleware/encode_uri_path.rb]: :+1: thanks for that, I was struggling to find something more appropriate

File: spec/posts_spec.rb (hunk 4 above)
Comment (toastercup) [inline spec/posts_spec.rb]: That's a hella handy tip, thanks @Casao!

**Comment by toastercup** [comment]
@Casao's feedback addressed - this is ready for review again.
